binr/r2r/r2r.c

```diff
@@ -532,15 +534,231 @@ static RThreadFunctionRet worker_th(RThread *th) {
+} R2RCharAlignment;
+
+typedef enum {
+	R2R_DIFF_MATCH, R2R_DIFF_DELETE, R2R_DIFF_INSERT
+} R2RPrintDiffMode;
+
 static void print_diff(const char *actual, const char *expected, bool diffchar) {
 	RDiff *d = r_diff_new ();
 #ifdef __WINDOWS__
 	d->diff_cmd = "git diff --no-index";
 #endif
+	size_t len_expected = strlen (expected);
+	size_t len_actual = strlen (actual);
 	if (diffchar) {
+		// Use Needleman–Wunsch to diffchar.
+		// This is an O(mn) algo in both space and time.
+		// Note that 64KB * 64KB * 2 = 8GB.
+		if (len_expected == len_actual) { // TODO Drop this condition
+			size_t dim = len_expected + 1;
+			st16 *align_table = malloc (dim * dim * sizeof (st16));
+			ut8 *align_expected = malloc (2 * len_expected);
+			ut8 *align_actual = malloc (2 * len_actual);
+			if (align_table && align_expected && align_actual) {
+				size_t row, col;
+				*align_table = 0;
+				for (row = 1; row < dim; row++) {
+					// TODO Clamping [ST16_MIN + 1, .]
+					*(align_table + row) = *(align_table + row * dim) = -(st16)row;
+				}
+				// Fill table
+				// TODO Cost-based instead of score-based system?
+				const st16 match = 1;
+				const st16 mismatch = -2;
+				const st16 gap = -1;
+				for (row = 1; row < dim; row++) {
+					for (col = 1; col < dim; col++) {
+						// TODO Clamping [ST16_MIN + 1, ST16_MAX]
+						st16 tl_score = *(align_table + (row - 1) * dim + col - 1)
+								+ (expected[col - 1] == actual[row - 1] ? match : mismatch);
+						st16 t_score = *(align_table + (row - 1) * dim + col) + gap;
+						st16 l_score = *(align_table + row * dim + col - 1) + gap;
+						st16 score;
+						if (tl_score >= t_score && tl_score >= l_score) {
+							score = tl_score;
+						} else if (t_score >= tl_score && t_score >= l_score) {
+							score = t_score;
+						} else {
+							score = l_score;
+						}
+						*(align_table + row * dim + col) = score;
+					}
+				}
+				// Print table (Debug)
+				char char_str[3] = { ' ' };
+				printf ("%4s ", char_str);
+				for (col = 0; col < dim; col++) {
+					if (col && expected[col - 1] == '\n') {
+						char_str[0] = '\\';
+						char_str[1] = 'n';
+					} else {
+						char_str[0] = !col ? ' ' : expected[col - 1];
+						char_str[1] = 0;
+					}
+					printf ("%4s ", char_str);
+				}
+				printf ("\n");
+				for (row = 0; row < dim; row++) {
+					if (row && actual[row - 1] == '\n') {
+						char_str[0] = '\\';
+						char_str[1] = 'n';
+					} else {
+						char_str[0] = !row ? ' ' : actual[row - 1];
+						char_str[1] = 0;
+					}
+					printf ("%4s ", char_str);
+					for (col = 0; col < dim; col++) {
+						printf ("%4d ", *(align_table + row * dim + col));
+					}
+					printf ("\n");
+				}
+				// Do alignment
+				size_t idx_expected = len_expected - 1;
+				size_t idx_actual = len_actual - 1;
+				size_t idx_align = 2 * len_expected - 1;
+				size_t pos_row = dim - 1;
+				size_t pos_col = dim - 1;
+				while (pos_row != 0 || pos_col != 0) {
```

Comment:
!=0 is implicit

Reply:
ok d4a002a7e921b5c58617b67e3709c71c82604751